File: sdks/python/src/verdocs/resources/templates.py
"""Template operations (js-sdk: Templates/Templates.ts).

The js-sdk applies some client-side legacy upgrades in getTemplate (copying
template_documents into documents, patching field settings). Those shims are
marked temporary in the js-sdk, so this SDK stays a faithful wire mirror and
returns responses exactly as the API sends them.

Template create speaks two encodings: JSON, or multipart when files are
attached. The js-sdk's own multipart encoding is broken (roles and fields
serialize to "[object Object]" and draw a 400), so the multipart path here
follows sdks/WIRE-NOTES.md instead: file parts all named "documents", string
fields only as text parts.
"""

from __future__ import annotations

import mimetypes
import os
from collections.abc import Sequence
from pathlib import Path
from typing import IO, TYPE_CHECKING, Any

import httpx

from ..errors import VerdocsConnectionError, api_error_from_response
from ..models import Template, TemplateCreateParams, TemplateList, TemplateListParams, TemplateUpdateParams
from ..models.templates import TemplateCreateFromSharepointParams

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_TEMPLATES_PATH = "/v2/templates"

# One attachment for a multipart create: a filesystem path, raw PDF bytes, an
# open binary file, or an httpx-style (filename, content[, content_type]) tuple.
TemplateFile = (
    str
    | os.PathLike[str]
    | bytes
    | bytearray
    | IO[bytes]
    | tuple[str, bytes | IO[bytes]]
    | tuple[str, bytes | IO[bytes], str]
)

# The create handler parses multipart text parts with the same zod schema as
# JSON bodies, and zod does not coerce strings, so only these string-typed
# fields survive the trip. Everything else must go via JSON or follow-up calls.
_MULTIPART_TEXT_FIELDS = ("name", "description", "visibility", "sender")

# The API gates uploads on each part's declared content type (PDF and DOCX
# only), and mimetypes does not know .docx on every platform, so we pin the
# two accepted types ourselves.
_MIME_BY_SUFFIX = {
    ".pdf": "application/pdf",
    ".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
}


def _guess_mime(filename: str) -> str:
    suffix = Path(filename).suffix.lower()
    if suffix in _MIME_BY_SUFFIX:
        return _MIME_BY_SUFFIX[suffix]
    guessed, _ = mimetypes.guess_type(filename)
    return guessed or "application/octet-stream"


def _file_part(entry: TemplateFile) -> Any:
    """Normalize one file input into something httpx can send as a file part.

    The server takes the document name from the part filename and gates the
    upload on the declared content type, so we fill in both wherever they can
    be derived. Tuples pass through untouched; the caller owns them.
    """
    if isinstance(entry, tuple):
        return entry
    if isinstance(entry, (str, os.PathLike)):
        path = Path(entry)
        return (path.name, path.read_bytes(), _guess_mime(path.name))
    if isinstance(entry, (bytes, bytearray)):
        # Bare bytes carry no filename, so we assume the common case. Pass a
        # (filename, content) tuple to control the name or to send DOCX.
        return ("document.pdf", bytes(entry), "application/pdf")
    name = getattr(entry, "name", None)
    if isinstance(name, str) and name:
        return (Path(name).name, entry, _guess_mime(name))
    return entry


def _document_parts(files: Sequence[TemplateFile]) -> list[tuple[str, Any]]:
    # The handler reads every file part from the one repeatable "documents" name.
    return [("documents", _file_part(entry)) for entry in files]


def _split_multipart_text(params: TemplateCreateParams) -> dict[str, str]:
    """Extract the text parts a multipart create can carry, or raise on fields that cannot ride."""
    body = params.model_dump(mode="json", exclude_unset=True)
    text: dict[str, str] = {}
    for key in _MULTIPART_TEXT_FIELDS:
        value = body.pop(key, None)
        if value is not None:
            text[key] = value
    if body:
        raise ValueError(
            "these create fields cannot be sent with file uploads: "
            + ", ".join(sorted(body))
            + ". Multipart text parts are string-only: send uri/data documents via a JSON create "
            "(no files argument), set reminders with update() after creating, and add roles with "
            "template_roles.create()."
        )
    return text


def _multipart_request(
    endpoint: VerdocsEndpoint, method: str, path: str, *, data: dict[str, str], files: list[tuple[str, Any]]
) -> httpx.Response:
    # endpoint._request only speaks query params and JSON bodies, so multipart
    # goes through the shared client with the same error mapping. Fold this
    # into _request if it ever grows files= support.
    try:
        response = endpoint._client.request(method, path, data=data, files=files)
    except httpx.TransportError as exc:
        raise VerdocsConnectionError(f"{method} {path} failed: {exc}") from exc

    if response.status_code >= 400:
        raise api_error_from_response(response)

    return response


async def _multipart_request_async(
    endpoint: AsyncVerdocsEndpoint, method: str, path: str, *, data: dict[str, str], files: list[tuple[str, Any]]
) -> httpx.Response:
    try:
        response = await endpoint._client.request(method, path, data=data, files=files)
    except httpx.TransportError as exc:
        raise VerdocsConnectionError(f"{method} {path} failed: {exc}") from exc

    if response.status_code >= 400:
        raise api_error_from_response(response)

    return response


def _list_query(params: TemplateListParams | None) -> dict[str, Any] | None:
    if params is None:
        return None
    return params.model_dump(mode="json", exclude_none=True)


def _write_body(
    params: TemplateCreateParams | TemplateCreateFromSharepointParams | TemplateUpdateParams,
) -> dict[str, Any]:
    # exclude_unset rather than exclude_none: an explicit None must reach the
    # wire (null disables reminders) while untouched fields stay off it.
    return params.model_dump(mode="json", exclude_unset=True)


class Templates:
    """Template calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def list(self, params: TemplateListParams | None = None) -> TemplateList:
        """Get the templates accessible to the caller, with optional filters.

        Calls GET /v2/templates. Mirrors js-sdk getTemplates.

        Example:
            page = endpoint.templates.list(TemplateListParams(visibility="private_shared", rows=10, page=0))

        Args:
            params: Optional filters, sorting, and pagination.

        Returns:
            One page of templates plus pagination counts.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", _TEMPLATES_PATH, params=_list_query(params))
        return TemplateList.model_validate(response.json())

    def get(self, template_id: str) -> Template:
        """Get one template by ID via GET /v2/templates/{template_id}.

        The caller needs at least View access to the template. Mirrors js-sdk getTemplate.

        Args:
            template_id: ID of the template to fetch.

        Returns:
            The requested template.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", f"{_TEMPLATES_PATH}/{template_id}")
        return Template.model_validate(response.json())

    def create(self, params: TemplateCreateParams, files: Sequence[TemplateFile] | None = None) -> Template:
        """Create a template via POST /v2/templates. Mirrors js-sdk createTemplate.

        Two encodings, chosen by the files argument. Without files the request
        is JSON and params may carry documents entries (uri or base64 data)
        plus inline roles. With files the request is multipart/form-data:
        every file rides as a part named "documents", only the string fields
        (name, description, visibility, sender) ride along as text parts, and
        setting anything else raises ValueError before a request is made. The
        API accepts PDF and DOCX only, judged by each part's declared content
        type.

        Inline fields are not supported on either path: the deployed API
        validates a fields array but never creates the rows, so add fields
        with endpoint.template_fields.create() afterwards.

        Example:
            template = endpoint.templates.create(
                TemplateCreateParams(name="NDA", visibility="shared"),
                files=["/contracts/nda.pdf"],
            )

        Args:
            params: Fields for the new template; only name is required.
            files: Optional documents to attach. Each entry may be a path, raw
                PDF bytes, an open binary file, or an httpx-style
                (filename, content[, content_type]) tuple. Bare bytes are sent
                as "document.pdf" with an application/pdf content type; use a
                tuple to control the name or to send DOCX bytes.

        Returns:
            The newly created template, with documents, fields, and roles.

        Raises:
            ValueError: files was given together with fields that cannot ride multipart.
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        if files:
            response = _multipart_request(
                self._endpoint,
                "POST",
                _TEMPLATES_PATH,
                data=_split_multipart_text(params),
                files=_document_parts(files),
            )
        else:
            response = self._endpoint._request("POST", _TEMPLATES_PATH, json=_write_body(params))
        return Template.model_validate(response.json())

    def duplicate(self, template_id: str, name: str) -> Template:
        """Duplicate a template via PUT /v2/templates/{template_id}. Mirrors js-sdk duplicateTemplate.

        Creates a complete clone, including settings (reminders and friends),
        fields, roles, and documents.

        Args:
            template_id: ID of the template to copy.
            name: Name for the new copy.

        Returns:
            The newly copied template.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request(
            "PUT", f"{_TEMPLATES_PATH}/{template_id}", json={"action": "duplicate", "name": name}
        )
        return Template.model_validate(response.json())

    def create_from_sharepoint(self, params: TemplateCreateFromSharepointParams) -> Template:
        """Create a template from a Sharepoint asset. Mirrors js-sdk createTemplateFromSharepoint.

        Dead on the deployed API: POST /v2/templates/from-sharepoint has no
        handler anywhere in the API, so every call fails. Ported for js-sdk
        shape parity only; keep it out of examples until the API ships the
        route.

        Args:
            params: Sharepoint site/item coordinates and the On-Behalf-Of token.

        Returns:
            The newly created template, if the API ever ships the route.

        Raises:
            VerdocsAPIError: Always today; the deployed API has no such route.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("POST", f"{_TEMPLATES_PATH}/from-sharepoint", json=_write_body(params))
        return Template.model_validate(response.json())

    def update(self, template_id: str, params: TemplateUpdateParams) -> Template:
        """Update a template via PATCH /v2/templates/{template_id}. Mirrors js-sdk updateTemplate.

        Args:
            template_id: ID of the template to update.
            params: The fields to change; unset fields are left alone.

        Returns:
            The updated template.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("PATCH", f"{_TEMPLATES_PATH}/{template_id}", json=_write_body(params))
        return Template.model_validate(response.json())

    def delete(self, template_id: str) -> None:
        """Delete a template via DELETE /v2/templates/{template_id}. Mirrors js-sdk deleteTemplate.

        The API answers with a bare success string that nothing consumes, so
        this returns None and relies on exceptions for failure.

        Args:
            template_id: ID of the template to delete.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        self._endpoint._request("DELETE", f"{_TEMPLATES_PATH}/{template_id}")

    def toggle_star(self, template_id: str) -> Template:
        """Toggle the caller's star on a template. Mirrors js-sdk toggleTemplateStar.

        Broken on both sides today: this posts the js-sdk's
        /v2/templates/{template_id}/stars/toggle path, which has no server
        route, and the server's own star route is a GET whose handler cannot
        be satisfied by any client. The deployed API rejects every call;
        retirement is pending. Ported for shape parity and excluded from
        conformance.

        Args:
            template_id: ID of the template to star or unstar.

        Returns:
            The updated template, if the API ever repairs the route.

        Raises:
            VerdocsAPIError: Always today; the deployed API has no such route.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("POST", f"{_TEMPLATES_PATH}/{template_id}/stars/toggle")
        return Template.model_validate(response.json())


class AsyncTemplates:
    """Template calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def list(self, params: TemplateListParams | None = None) -> TemplateList:
        """Get the templates accessible to the caller, with optional filters.

        Calls GET /v2/templates. Mirrors js-sdk getTemplates.

        Example:
            page = await endpoint.templates.list(TemplateListParams(visibility="private_shared", rows=10, page=0))

        Args:
            params: Optional filters, sorting, and pagination.

        Returns:
            One page of templates plus pagination counts.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _TEMPLATES_PATH, params=_list_query(params))
        return TemplateList.model_validate(response.json())

    async def get(self, template_id: str) -> Template:
        """Get one template by ID via GET /v2/templates/{template_id}.

        The caller needs at least View access to the template. Mirrors js-sdk getTemplate.

        Args:
            template_id: ID of the template to fetch.

        Returns:
            The requested template.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", f"{_TEMPLATES_PATH}/{template_id}")
        return Template.model_validate(response.json())

    async def create(self, params: TemplateCreateParams, files: Sequence[TemplateFile] | None = None) -> Template:
        """Create a template via POST /v2/templates. Mirrors js-sdk createTemplate.

        Two encodings, chosen by the files argument. Without files the request
        is JSON and params may carry documents entries (uri or base64 data)
        plus inline roles. With files the request is multipart/form-data:
        every file rides as a part named "documents", only the string fields
        (name, description, visibility, sender) ride along as text parts, and
        setting anything else raises ValueError before a request is made. The
        API accepts PDF and DOCX only, judged by each part's declared content
        type.

        Inline fields are not supported on either path: the deployed API
        validates a fields array but never creates the rows, so add fields
        with endpoint.template_fields.create() afterwards.

        Example:
            template = await endpoint.templates.create(
                TemplateCreateParams(name="NDA", visibility="shared"),
                files=["/contracts/nda.pdf"],
            )

        Args:
            params: Fields for the new template; only name is required.
            files: Optional documents to attach. Each entry may be a path, raw
                PDF bytes, an open binary file, or an httpx-style
                (filename, content[, content_type]) tuple. Bare bytes are sent
                as "document.pdf" with an application/pdf content type; use a
                tuple to control the name or to send DOCX bytes.

        Returns:
            The newly created template, with documents, fields, and roles.

        Raises:
            ValueError: files was given together with fields that cannot ride multipart.
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        if files:
            response = await _multipart_request_async(
                self._endpoint,
                "POST",
                _TEMPLATES_PATH,
                data=_split_multipart_text(params),
                files=_document_parts(files),
            )
        else:
            response = await self._endpoint._request("POST", _TEMPLATES_PATH, json=_write_body(params))
        return Template.model_validate(response.json())

    async def duplicate(self, template_id: str, name: str) -> Template:
        """Duplicate a template via PUT /v2/templates/{template_id}. Mirrors js-sdk duplicateTemplate.

        Creates a complete clone, including settings (reminders and friends),
        fields, roles, and documents.

        Args:
            template_id: ID of the template to copy.
            name: Name for the new copy.

        Returns:
            The newly copied template.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request(
            "PUT", f"{_TEMPLATES_PATH}/{template_id}", json={"action": "duplicate", "name": name}
        )
        return Template.model_validate(response.json())

    async def create_from_sharepoint(self, params: TemplateCreateFromSharepointParams) -> Template:
        """Create a template from a Sharepoint asset. Mirrors js-sdk createTemplateFromSharepoint.

        Dead on the deployed API: POST /v2/templates/from-sharepoint has no
        handler anywhere in the API, so every call fails. Ported for js-sdk
        shape parity only; keep it out of examples until the API ships the
        route.

        Args:
            params: Sharepoint site/item coordinates and the On-Behalf-Of token.

        Returns:
            The newly created template, if the API ever ships the route.

        Raises:
            VerdocsAPIError: Always today; the deployed API has no such route.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", f"{_TEMPLATES_PATH}/from-sharepoint", json=_write_body(params))
        return Template.model_validate(response.json())

    async def update(self, template_id: str, params: TemplateUpdateParams) -> Template:
        """Update a template via PATCH /v2/templates/{template_id}. Mirrors js-sdk updateTemplate.

        Args:
            template_id: ID of the template to update.
            params: The fields to change; unset fields are left alone.

        Returns:
            The updated template.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PATCH", f"{_TEMPLATES_PATH}/{template_id}", json=_write_body(params))
        return Template.model_validate(response.json())

    async def delete(self, template_id: str) -> None:
        """Delete a template via DELETE /v2/templates/{template_id}. Mirrors js-sdk deleteTemplate.

        The API answers with a bare success string that nothing consumes, so
        this returns None and relies on exceptions for failure.

        Args:
            template_id: ID of the template to delete.

        Raises:
            NotFoundError: No visible template has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        await self._endpoint._request("DELETE", f"{_TEMPLATES_PATH}/{template_id}")

    async def toggle_star(self, template_id: str) -> Template:
        """Toggle the caller's star on a template. Mirrors js-sdk toggleTemplateStar.

        Broken on both sides today: this posts the js-sdk's
        /v2/templates/{template_id}/stars/toggle path, which has no server
        route, and the server's own star route is a GET whose handler cannot
        be satisfied by any client. The deployed API rejects every call;
        retirement is pending. Ported for shape parity and excluded from
        conformance.

        Args:
            template_id: ID of the template to star or unstar.

        Returns:
            The updated template, if the API ever repairs the route.

        Raises:
            VerdocsAPIError: Always today; the deployed API has no such route.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", f"{_TEMPLATES_PATH}/{template_id}/stars/toggle")
        return Template.model_validate(response.json())

File: sdks/csharp/src/Verdocs.Sdk/Resources/Templates.cs
using System.Globalization;
using System.Net.Http.Headers;
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Template calls, reached through <see cref="VerdocsEndpoint.Templates"/>.
/// </summary>
public sealed class Templates
{
    private readonly VerdocsEndpoint _endpoint;

    internal Templates(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Gets the templates accessible to the caller, with optional filters.
    ///
    /// <example>
    /// <code>
    /// var page = await endpoint.Templates.ListAsync(new GetTemplatesOptions
    /// {
    ///     Visibility = TemplateVisibilityFilter.PrivateShared,
    ///     Rows = 10,
    ///     Page = 0,
    /// });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="options">Optional filters, sorting, and paging.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>One page of templates plus paging counts.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the session is invalid.</exception>
    /// <sdkOperation>template.getTemplates</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<TemplateList> ListAsync(GetTemplatesOptions? options = null, CancellationToken cancellationToken = default)
    {
        return _endpoint.SendAsync<TemplateList>(HttpMethod.Get, BuildTemplatesPath(options), null, cancellationToken);
    }

    /// <summary>
    /// Gets one template by its ID. The caller must have at least view access to it. The
    /// detail response includes the template's roles, documents, and fields, which the list
    /// response omits.
    ///
    /// <example>
    /// <code>
    /// var template = await endpoint.Templates.GetAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The requested template.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the template was not found.</exception>
    /// <sdkOperation>template.getTemplate</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Template> GetAsync(string templateId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        return _endpoint.SendAsync<Template>(HttpMethod.Get, "/v2/templates/" + Uri.EscapeDataString(templateId), null, cancellationToken);
    }

    /// <summary>
    /// Creates a template, sending the request as JSON. Documents may be attached by URI or
    /// Base64 data via <see cref="CreateTemplateRequest.Documents"/>; to upload local files
    /// directly, use the overload that takes <see cref="TemplateFileUpload"/> parts. Roles
    /// are created inline when supplied. Fields cannot be created here (the server validates
    /// a create-time fields array but discards it); add them afterward through
    /// <see cref="TemplateFields.CreateAsync"/>.
    ///
    /// <example>
    /// <code>
    /// var template = await endpoint.Templates.CreateAsync(new CreateTemplateRequest
    /// {
    ///     Name = "Bill of Sale",
    ///     Visibility = TemplateVisibility.Shared,
    /// });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">The template to create.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new template, including its documents, roles, and fields.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because a document was an unsupported type.</exception>
    /// <sdkOperation>template.createTemplate</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    /// <sdkGettingStarted />
    public Task<Template> CreateAsync(CreateTemplateRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<Template>(HttpMethod.Post, "/v2/templates", request, cancellationToken);
    }

    /// <summary>
    /// Creates a template and uploads local files in the same call, as multipart form data
    /// with every file part named "documents". Multipart text parts carry strings only, so
    /// this path sends just the name, description, visibility, and sender; a request that
    /// sets roles, document sources, or reminder settings is rejected here because those
    /// cannot ride multipart (the js-sdk tries and the server rejects the result). Create
    /// with roles and reminders via the JSON overload, or add them after this call through
    /// <see cref="TemplateRoles.CreateAsync"/> and <see cref="UpdateAsync"/>.
    /// </summary>
    /// <param name="request">The template to create. Only string settings ride along with files.</param>
    /// <param name="documents">Files to upload; at least one. PDF or DOCX, up to 25 MB each.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new template, including its documents, roles, and fields.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because a document was an unsupported type.</exception>
    public Task<Template> CreateAsync(
        CreateTemplateRequest request,
        IReadOnlyCollection<TemplateFileUpload> documents,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        ArgumentNullException.ThrowIfNull(documents);
        if (documents.Count == 0)
        {
            throw new ArgumentException(
                "At least one document is required on the multipart path; use the JSON overload when there are no files to upload.",
                nameof(documents));
        }

        if (request.Roles is not null || request.Documents is not null
            || request.InitialReminder is not null || request.FollowupReminders is not null || request.MaxReminderDays is not null)
        {
            throw new ArgumentException(
                "Roles, document sources, and reminder settings cannot be sent with file uploads; multipart text parts carry strings only. "
                + "Use the JSON overload, or add roles and reminders after creation.",
                nameof(request));
        }

        var content = new MultipartFormDataContent();
        AddTextPart(content, "name", request.Name);
        AddTextPart(content, "description", request.Description);
        AddTextPart(content, "visibility", request.Visibility);
        AddTextPart(content, "sender", request.Sender);

        foreach (var document in documents)
        {
            content.Add(CreateFilePart(document), "\"documents\"", "\"" + document.FileName + "\"");
        }

        return _endpoint.SendAsync<Template>(HttpMethod.Post, "/v2/templates", content, cancellationToken);
    }

    /// <summary>
    /// Updates a template's settings. Reminder settings are rewritten on every update; see
    /// <see cref="UpdateTemplateRequest"/> for the details.
    ///
    /// <example>
    /// <code>
    /// var template = await endpoint.Templates.UpdateAsync(
    ///     "d2338742-f3a1-465b-8592-806587413cc1",
    ///     new UpdateTemplateRequest { Name = "Bill of Sale (v2)" });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template's unique ID.</param>
    /// <param name="request">The settings to change.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The updated template.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller may not edit the template.</exception>
    /// <sdkOperation>template.updateTemplate</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Template> UpdateAsync(string templateId, UpdateTemplateRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<Template>(
            HttpMethod.Patch, "/v2/templates/" + Uri.EscapeDataString(templateId), request, cancellationToken);
    }

    /// <summary>
    /// Deletes a template.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Templates.DeleteAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the template has been deleted.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller may not edit the template.</exception>
    /// <sdkOperation>template.deleteTemplate</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task DeleteAsync(string templateId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        return _endpoint.SendVoidAsync(
            HttpMethod.Delete, "/v2/templates/" + Uri.EscapeDataString(templateId), null, cancellationToken);
    }

    /// <summary>
    /// Duplicates a template as a complete clone: settings (reminders included), documents,
    /// roles, and fields.
    ///
    /// <example>
    /// <code>
    /// var copy = await endpoint.Templates.DuplicateAsync("d2338742-f3a1-465b-8592-806587413cc1", "Bill of Sale (copy)");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template to copy.</param>
    /// <param name="name">Name for the new copy.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new copy.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the template was not found.</exception>
    /// <sdkOperation>template.duplicateTemplate</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Template> DuplicateAsync(string templateId, string name, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        ArgumentException.ThrowIfNullOrEmpty(name);
        return _endpoint.SendAsync<Template>(
            HttpMethod.Put, "/v2/templates/" + Uri.EscapeDataString(templateId), new { action = "duplicate", name }, cancellationToken);
    }

    /// <summary>
    /// Creates a template from a Sharepoint file. Dead on the deployed API: no handler exists
    /// for this route, so every call fails. Ported to match the js-sdk surface; retirement is
    /// pending.
    ///
    /// <example>
    /// <code>
    /// var template = await endpoint.Templates.CreateFromSharepointAsync(new CreateTemplateFromSharepointRequest
    /// {
    ///     Name = "Bill of Sale",
    ///     SiteId = "site-id",
    ///     ItemId = "item-id",
    ///     OboToken = "obo-token",
    /// });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">The Sharepoint source and template name.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new template, if the API ever serves this route.</returns>
    /// <exception cref="VerdocsApiException">Always thrown today; the deployed API has no handler for this call.</exception>
    /// <sdkOperation>template.createTemplateFromSharepoint</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Template> CreateFromSharepointAsync(CreateTemplateFromSharepointRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<Template>(HttpMethod.Post, "/v2/templates/from-sharepoint", request, cancellationToken);
    }

    /// <summary>
    /// Toggles the caller's star on a template. Broken on both sides today: the js-sdk (and
    /// this port, which mirrors it) POSTs a path that does not exist, and the server's own
    /// star route is a GET whose handler cannot be satisfied, so the deployed API rejects
    /// every client. Retirement is pending an API fix.
    ///
    /// <example>
    /// <code>
    /// var template = await endpoint.Templates.ToggleStarAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template to star or unstar.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The template, if the API ever serves this route.</returns>
    /// <exception cref="VerdocsApiException">Always thrown today; the deployed API rejects this call.</exception>
    /// <sdkOperation>template.toggleTemplateStar</sdkOperation>
    /// <sdkGroup>Template</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Template> ToggleStarAsync(string templateId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        return _endpoint.SendAsync<Template>(
            HttpMethod.Post, "/v2/templates/" + Uri.EscapeDataString(templateId) + "/stars/toggle", null, cancellationToken);
    }

    private static void AddTextPart(MultipartFormDataContent content, string name, string? value)
    {
        if (value is null)
        {
            return;
        }

        // Names are pre-quoted because .NET only quotes multipart part names when it must,
        // while browsers (and so the API's usual traffic) always quote them.
        content.Add(new StringContent(value), "\"" + name + "\"");
    }

    private static StreamContent CreateFilePart(TemplateFileUpload document)
    {
        // The server runs its PDF/DOCX check against the declared part content type, so it
        // has to be set per part rather than left for the transport to default.
        var part = new StreamContent(document.Content);
        part.Headers.ContentType = new MediaTypeHeaderValue(document.ContentType);
        return part;
    }

    private static string BuildTemplatesPath(GetTemplatesOptions? options)
    {
        const string path = "/v2/templates";
        if (options is null)
        {
            return path;
        }

        var parameters = new List<string>();

        void Add(string key, string value) => parameters.Add(key + "=" + Uri.EscapeDataString(value));

        if (options.Q is { } q)
        {
            Add("q", q);
        }

        if (options.IsStarred is { } isStarred)
        {
            Add("is_starred", isStarred ? "true" : "false");
        }

        if (options.IsCreator is { } isCreator)
        {
            Add("is_creator", isCreator ? "true" : "false");
        }

        if (options.Visibility is { } visibility)
        {
            Add("visibility", ToWireValue(visibility));
        }

        if (options.SortBy is { } sortBy)
        {
            Add("sort_by", ToWireValue(sortBy));
        }

        if (options.Ascending is { } ascending)
        {
            Add("ascending", ascending ? "true" : "false");
        }

        if (options.Rows is { } rows)
        {
            Add("rows", rows.ToString(CultureInfo.InvariantCulture));
        }

        if (options.Page is { } page)
        {
            Add("page", page.ToString(CultureInfo.InvariantCulture));
        }

        return parameters.Count == 0 ? path : path + "?" + string.Join("&", parameters);
    }

    private static string ToWireValue(TemplateVisibilityFilter visibility) => visibility switch
    {
        TemplateVisibilityFilter.PrivateShared => "private_shared",
        TemplateVisibilityFilter.Private => "private",
        TemplateVisibilityFilter.Shared => "shared",
        TemplateVisibilityFilter.Public => "public",
        _ => throw new ArgumentOutOfRangeException(nameof(visibility)),
    };

    private static string ToWireValue(TemplateSortBy sortBy) => sortBy switch
    {
        TemplateSortBy.CreatedAt => "created_at",
        TemplateSortBy.UpdatedAt => "updated_at",
        TemplateSortBy.Name => "name",
        TemplateSortBy.LastUsedAt => "last_used_at",
        TemplateSortBy.Counter => "counter",
        TemplateSortBy.StarCounter => "star_counter",
        _ => throw new ArgumentOutOfRangeException(nameof(sortBy)),
    };
}
